### PR TITLE
[JS] Add a new OptimizeForJS pass

### PR DESCRIPTION
**Pull request by MaxGraey**
Add new OptimizeForJS pass which contain some rewriting rules specific to JavaScript.

LLVM usually lower `x != 0 && (x & (x - 1)) == 0` (isPowerOf2) to `popcnt(x) == 1` which is ok for wasm and other targets but quite expensive for JavaScript. In this PR we lower `popcnt` pattern back to `isPowerOf2` pattern.

Optimized version of `x != 0 && (x & (x - 1)) == 0`:
```ts
!(!x | (x & (x - 1))
```

### CODE REVIEW

**Comment by MaxGraey** [comment]
Hmm, it looks like `visitUnary` is executed before `visitBinary`. so we cannot map the replaced function `__wasm_popcnt_i32` but this would not be optimal. I would like to rewrite the rule earlier. 

@tlively @kripken What would you advise? Split into two passes? Create a PreWalker or is there an easier solution?

**Comment by kripken** [comment]
Hmm, yes, the order means that `popcount(x) == 1` will see the `==`  last, which is too late if we've already lowered the `popcount`. But, in any case I think another pass would be a good idea, yeah - `RemoveNonJSOps` should stay focused on lowering, not optimization. See `AvoidReinterprets` which is run by `wasm2js.h` to do one form of JS-only optimization. So I'd suggest adding a new pre pass, as you said, that would be run before `remove-non-js-ops`. Perhaps `pre-js-ops`?

**Comment by MaxGraey** [comment]
> Perhaps pre-js-ops?

Yeah. Sounds good to me

**Comment by MaxGraey** [comment]
Hmm, I'm honestly at a loss to understand what's wrong?

```
[wasm-validator error in function is-power-of-2] expression seen more than once in the tree in is-power-of-2 on 0x555556176a58
```

File: src/passes/PreJSRewriter.cpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+#include <wasm.h>
+
+#include "wasm-builder.h"
+#include <ir/literal-utils.h>
+#include <ir/localize.h>
+
+namespace wasm {
+
+struct PreJSRewriterPass : public WalkerPass<PostWalker<PreJSRewriterPass>> {
+  std::unique_ptr<Builder> builder;
```
Comment (tlively) [inline src/passes/PreJSRewriter.cpp:27]: Rather than have a `unique_ptr<Builder>` as a class member, it would be slightly simpler just to construct a new `Builder` as a local variable in any method where it is needed.

File: src/passes/PreJSRewriter.cpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+  void visitBinary(Binary* curr) {
+    switch (curr->op) {
+      case EqInt32:
+      case EqInt64:
+        // Rewrite popcnt(x) == 1   ==>   !(!x | (x & (x - 1))
+        if (auto* lhs = curr->left->dynCast<Unary>()) {
+          if (auto* rhs = curr->right->dynCast<Const>()) {
+            if ((lhs->op == PopcntInt32 || lhs->op == PopcntInt64) &&
+                rhs->value.getInteger() == 1LL) {
```
Comment (tlively) [inline src/passes/PreJSRewriter.cpp:48]: This could use matchers from match.h to be more readable.

File: src/passes/PreJSRewriter.cpp [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+      case CopySignFloat64:
+        rewriteCopysign(curr);
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  void rewriteCopysign(Binary* curr) {
```
Comment (tlively) [inline src/passes/PreJSRewriter.cpp:65]: It would be good to have a comment showing what transformation is being done here, too.

**Review by tlively — COMMENTED** [review]
It would be good to add a test in test/lit/passes that uses wasm-opt to just run this new pass, rather than relying on wasm-to-JS tests for testing it.

**Comment by MaxGraey** [comment]
@tlively @kripken RemoveNonJSOps currently contain only lowered to functions / stubs methods. What do think rename `RemoveNonJSOps` to `JSLowering` or something like this?

File: src/wasm2js.h [hunk 4]
```diff
@@ -341,6 +341,9 @@ Ref Wasm2JSBuilder::processWasm(Module* wasm, Name funcName) {
     // TODO: only legalize if necessary - emscripten would already do so, and
     //       likely other toolchains. but spec test suite needs that.
     runner.add("legalize-js-interface");
+    // Before lowering non-JS operations we can rewrite some instructions which
+    // may simplify next passes
+    runner.add("pre-js-rewriter");
```
Comment (tlively) [inline src/wasm2js.h:346]: I would name the pass "pre-js-rewrite" to make it more of a verb instead of a noun.

File: test/passes/remove-non-js-ops.txt [hunk 5]
```diff
@@ -33,39 +33,15 @@
...
-    (i64.and
-     (i64.reinterpret_f64
-      (local.get $1)
-     )
-     (i64.const -9223372036854775808)
-    )
-   )
+  (f64.copysign
+   (local.get $0)
+   (local.get $1)
```
Comment (tlively) [inline test/passes/remove-non-js-ops.txt:38]: These test functions can probably be removed entirely now that they don't test anything interesting.

**Review by tlively — COMMENTED** [review]
I'll defer to @kripken on the pass renaming question.

File: test/passes/remove-non-js-ops.txt (hunk 5 above)
Comment (MaxGraey) [reply test/passes/remove-non-js-ops.txt:38]: yeah, I'll move this fn to new test

**Comment by MaxGraey** [comment]
@tlively all done!

**Comment by kripken** [comment]
I think we can rename passes where it makes sense. JSLowering might be more consistent with other existing passes. But let's consider that separately from this PR.

**Comment by MaxGraey** [comment]
Sure. This PR ready for review / merge

**Review by kripken — COMMENTED** [review]
IIUC, @MaxGraey , your idea is that PreJSRewriter rewrites the IR to more JS-friendly IR. And that RemoveNonJSOps replaces certain instructions with calls to intrinsics? And therefore there is code being moved from one to the other?

Another way to consider this is that the pre pass could be OptimizeForJS, a pass that rewrites the IR to be more optimizable for JS later. That pass would be optional - when running in -O0, we would not execute it. And RemoveNonJSOps would remain as it is now, which is to do all the necessary changes to allow JS execution.

In summary, IIUC, the first model is "one pass does IR transforms, the other does instruction => intrinsic calls", and the second model is "one pass does optional optimization, the other does mandatory changes".

Is that a fair summary?

**Comment by MaxGraey** [comment]
> your idea is that PreJSRewriter rewrites the IR to more JS-friendly IR. And that RemoveNonJSOps replaces certain instructions with calls to intrinsics? And therefore there is code being moved from one to the other?

yes!

> Another way to consider this is that the pre pass could be OptimizeForJS, a pass that rewrites the IR to be more optimizable for JS later. That pass would be optional - when running in -O0, we would not execute it. And RemoveNonJSOps would remain as it is now, which is to do all the necessary changes to allow JS execution.

This option works for me, too.

So the decision is yours. I'll take either option. Do you find the second option more attractive?

**Comment by kripken** [comment]
I don't have strong feelings myself. But I lean to option 2, since the mandatory/just-an-optimization distinction is more significant, I think. It's nice to not run optional opts in debug builds for various reasons.

**Comment by MaxGraey** [comment]
Done!

File: src/passes/RemoveNonJSOps.cpp [hunk 6]
```diff
@@ -324,6 +285,57 @@ struct RemoveNonJSOpsPass : public WalkerPass<PostWalker<RemoveNonJSOpsPass>> {
   void visitGlobalGet(GlobalGet* curr) {
     neededImportedGlobals.insert(std::make_pair(curr->name, curr->type));
   }
+
+  void rewriteCopysign(Binary* curr) {
```
Comment (tlively) [inline src/passes/RemoveNonJSOps.cpp:289]: Unless there was a good reason to move it, it would be good to put this function back where it was to keep the diff smaller. (It would be good to keep the new comment, though.)

**Review by tlively — APPROVED** [review]
LGTM with that last cleanup!

File: src/passes/RemoveNonJSOps.cpp (hunk 6 above)
Comment (MaxGraey) [reply src/passes/RemoveNonJSOps.cpp:289]: Ah, sure!

File: src/passes/OptimizeForJS.cpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      case Type::i64:
+        eqzOp = EqZInt64;
+        andOp = AndInt64;
+        subOp = SubInt64;
+        litOne = Literal::makeOne(Type::i64);
+        break;
+
+      default:
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (MaxGraey) [reply src/passes/OptimizeForJS.cpp:69]: Yeah, that's will be shorter
